README.md
# The pylibcklb library 

[![build status](https://gitlab.ecklebe.de/open-source/python/pylibcklb/badges/master/build.svg)](https://gitlab.ecklebe.de/open-source/python/pylibcklb/commits/master) (mirroring, build and install on Linux)

The pylibcklb library an package of different functions and classes for programming in python.

## Requirements

The package was programmed with **python 3.6 64 bit**. Older or newer versions of python were not been tested.

## Installation

To install the package to the python packages you need **pip** the package install management which is included in the python interpreter.

 1. Call the command line under windows as admin
 2. Type the following to command line and that's is

```batch
pip install https://github.com/Ecklebe/pylibcklb/archive/master.zip
```

## Usage
To use (with caution), simply use this small example:

Write main.py:

```python
# Import of the package functions
from pylibcklb.FunctionLibrary import HelloWorld

# Main fucntion to call with python interpreter
def main():

    # Simple function call
    HelloWorld()

if __name__ == "__main__":
    main()
```

Call the main.py with python from command line:
```batch
python.exe main.py
```
Output of the command line: 

    >>> Hello world i am the pylibcklb package
	
## Changes
For the changes of each version see the [CHANGELOG.md][CHANGELOG] in this folder.

[CHANGELOG]: ./CHANGELOG.md



CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).


## [Unreleased]

## [1.2.0] - 2018-02-23
### Added 
- Add functions to work with OpenCV 3.4

## [1.1.0] - 2018-02-09
### Added 
- New function to check for prefix in text and remove if existing
- New function to change the keyname of an xml item
- New fuction to update xml key value pair in the attributes
- New fuction to create executable with pyinstaller (works!)

## [1.0.0] - 2017-12-11
### Configured 
- restructure of pylibcklb package 

## [0.9.0] - 2017-12-10
### Added 
- add function to get relative path to target path from current working path
- add function to print message in green and red for success or error
### Fixed
- Make CreateDir and CreateFile more safety through call submethods in try-except-block

## [0.8.0] - 2017-11-29
### Added
- Rezise start dialog and info dialog to 0.5 of the display size
- Add method to create menu bar and the entry 'Edit'
- Add classes to create main window without anything else and with pre created stuff

## [0.7.0] - 2017-11-25
### Added
- add function to save changes of dict back to file to make them static
- add checkbox to info dialog and start dialog

## [0.6.0] - 2017-11-23
### Added
- add new dialog classes for startscreen dialog and info dialog

### Fixed
- There are some fixes to the previous version for import bugs and other things

## [0.5.0] - 2017-11-22
### Added
- new class to get thread for searching specific filetypes in directory
- add new classes to handle toolbars and undo and redo in the toolbar 
- add function to set drag drop mode of an qtreeview element

## [0.4.0] - 2017-11-14
### Added
- Add first work on script library for installer like pyinstaller

## [0.3.0] - 2017-11-10
### Added
- Add LGPLv3 License 

## [0.2.0] - 2017-11-07
### Added
- Add new functions to create a file and and directory
- Add new comments to save and load folder function
- Add new functions to create message boxes for error, warning, information and questions

## 0.1.0 - 2017-06-02
### Added
- Init commit and push of different elements to new repository

[Unreleased]: https://gitlab.ecklebe.de/open-source/pylibcklb/compare/v1.2.0...master
[1.2.0]: https://gitlab.ecklebe.de/open-source/pylibcklb/compare/v1.1.0...v1.2.0
[1.1.0]: https://gitlab.ecklebe.de/open-source/pylibcklb/compare/v1.0.0...v1.1.0
[1.0.0]: https://gitlab.ecklebe.de/open-source/pylibcklb/compare/v0.9.0...v1.0.0
[0.9.0]: https://gitlab.ecklebe.de/open-source/pylibcklb/compare/v0.8.0...v0.9.0
[0.8.0]: https://gitlab.ecklebe.de/open-source/pylibcklb/compare/v0.7.0...v0.8.0
[0.7.0]: https://gitlab.ecklebe.de/open-source/pylibcklb/compare/v0.6.0...v0.7.0
[0.6.0]: https://gitlab.ecklebe.de/open-source/pylibcklb/compare/v0.5.0...v0.6.0
[0.5.0]: https://gitlab.ecklebe.de/open-source/pylibcklb/compare/v0.4.0...v0.5.0
[0.4.0]: https://gitlab.ecklebe.de/open-source/pylibcklb/compare/v0.3.0...v0.4.0
[0.3.0]: https://gitlab.ecklebe.de/open-source/pylibcklb/compare/v0.2.0...v0.3.0
[0.2.0]: https://gitlab.ecklebe.de/open-source/pylibcklb/compare/v0.1.0...v0.2.0

pylibcklb/metadata.py
## Metadata file for the pylibcklb package
#
# @file		    metadata.py
# @author	    Tobias Ecklebe
# @date		    09.11.2017
# @version	    0.1.0
# @note		    The metadata is stored here. It can be used by any other module in this project this way
# 
# @pre          The library was developed with python 3.6 64 bit
#
# @bug          No bugs at the moment.
#
# @warning      No warnings at the moment. 
#
# @copyright    pylibcklb package
#               Copyright (C) 2017  Tobias Ecklebe
#
#               This program is free software: you can redistribute it and/or modify
#               it under the terms of the GNU Lesser General Public License as published by
#               the Free Software Foundation, either version 3 of the License, or
#               (at your option) any later version.
#
#               This program is distributed in the hope that it will be useful,
#               but WITHOUT ANY WARRANTY; without even the implied warranty of
#               MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#               GNU Lesser General Public License for more details.
#
#               You should have received a copy of the GNU Lesser General Public License
#               along with this program.  If not, see <http://www.gnu.org/licenses/>.
#

"""
The metadata is stored here. It can be used by any other module in this project this way
"""
from pylibcklb.version import __version__
from pylibcklb.ClassLibrary import cDebug 

class PackageVariables:

    DebugLevel = cDebug.LEVEL_ZERO

class GitRepository:
    """
    Class that stores information about the git repository sites used by this project
    """

    repository_name = "pylibcklb"
    """
    The name of the repository
    """

    gitlab_owner = "open-source/python"
    """
    The project's owner's username on Gitlab
    """

    gitlab_site_url = "https://gitlab.ecklebe.de/"
    """
    The address of the Gitlab instance
    """

    gitlab_url = gitlab_site_url + gitlab_owner + "/" + repository_name
    """
    The Gitlab Project URL
    """


class General:
    """
    Class that stores general information about a project
    """

    project_description = 'The pylibcklb is a library of functions and classes created from me and were used in different other projects of me.'
    """
    A short description of the project
    """

    version_number = __version__
    """
    The current version of the program.
    """

    author_names = "Tobias Ecklebe"
    """
    The name(s) of the project author(s)
    """

    author_emails = "tobias.ecklebe@outlook.de"
    """
    The email address(es) of the project author(s)
    """

    license_type = "GNU LGPLv3"
    """
    The project's license type
    """

    project_name = GitRepository.repository_name
    """
    The name of the project
    """

    download_master_zip = GitRepository.gitlab_url + "/repository/archive.zip?ref=master"
    """
    A URL linking to the current source zip file of the master branch.
    """


class Variables:
    """
    Variables used for distributing with setuptools to the python package index
    """

    classifiers = [

        "Natural Language :: English",
        'Programming Language :: Python :: 3.6',
        "License :: OSI Approved :: GNU Lesser General Public License v3 (LGPLv3)",
        "Operating System :: Microsoft :: Windows :: Windows 10",
        "Topic :: Software Development :: Libraries"

    ]
    """
    The list trove classifiers applicable to this project
    """

    install_requires = ["lxml", "markdown2", "numpy", "pyinstaller", "colorama"]
    """
    Python Packaging Index dependencies
    """

    extras_require = {"gui": ["PyQt5"]}
    """
    Optional dependencies for Pypi
    """

    test_require = ["nose"]
    """
    Optional dependencies for test
    """

    name = General.project_name
    """
    The name of the project on Pypi
    """

    version = General.version_number
    """
    The version of the project on pypi
    """

    description = General.project_description
    """
    The short description of the project on pypi
    """

    url = GitRepository.gitlab_url
    """
    A URL linking to the home page of the project, in this case a
    self-hosted Gitlab page
    """

    download_url = General.download_master_zip
    """
    A link to the current source zip of the project
    """

    author = General.author_names
    """
    The author(s) of this project
    """

    author_email = General.author_emails
    """
    The email adress(es) of the author(s)
    """

    license = General.license_type
    """
    The License used in this project
    """